source/server/hot_restart_impl.cc

```diff
@@ -110,12 +110,16 @@ Stats::RawStatData* HotRestartImpl::alloc(const std::string& name) {
   // Try to find the existing slot in shared memory, otherwise allocate a new one.
   std::unique_lock<Thread::BasicLockable> lock(stat_lock_);
   for (Stats::RawStatData& data : shmem_.stats_slots_) {
+    if (data.initialized() && data.matches(name)) {
```

Comment:
perf nit: I think you can still do this in 1 pass by keeping track of the first empty entry while you scan the entire range. Then if you don't find it, you immediately know where to put it or if you are OOM.  

It occurs to me also that at startup we could built free list and a map from shared memory and it would all be O(1), but I probably wouldn't bother right now. Feel free to put in a TODO along those lines.

Reply:
I came up with the same ideas when I thought about this over the weekend.  

I came across this bug in the context of making the number of stats tunable, and I was thinking about the O(n^2) complexity of allocating all the stats when N is larger.  I'll keep discussions of how to make this substantially better over there, but I'll do this in one loop here.